src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs

```diff
@@ -193,5 +194,17 @@ private static bool SeekBackwardsAndRead(Stream stream, byte[] buffer, out int b
                 return true;
             }
         }
+
+        // Converts the specified string into bytes using the optional specified encoding. If null, then the encoding is calculated from the string itself.
+        internal static byte[] EncodeStringToBytes(string text, Encoding? encoding, out bool isUTF8)
+        {
+            Debug.Assert(text != null);
+
+            if (encoding == null)
+                encoding = RequiresUnicode(text) ? Encoding.UTF8 : Encoding.ASCII;
+
+            isUTF8 = encoding.Equals(Encoding.UTF8);
```

Comment:
This logic is incorrect, as `Encoding.Equals` relies on the particular arguments passed to the ctor. For example, `new Utf8Encoding(false, false).Equals(new Utf8Encoding(true, true))` returns _false_.

Recommended way to check "is this UTF-8?" is:
```cs
isUTF8 = (encoding.CodePage == 65001);
```

This pattern is similar to the _CodePage_ pattern check mentioned for UTF-7 [here](https://docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/5.0/utf-7-code-paths-obsolete#recommended-action).